File: skills/plotting/tests/test_density.R
#!/usr/bin/env Rscript
# Tests for pavlab_density (R/ggplot2 backend).
# Run from skills/plotting/:
#   Rscript tests/test_density.R

if (!requireNamespace("ggplot2", quietly = TRUE))
  stop("ggplot2 required: install.packages('ggplot2')")

source("R/palettes.R")
source("R/pavlab_scatter.R")
source("R/pavlab_density.R")

cat("test_density.R\n")
cat("==============\n")

# ---- Helpers -----------------------------------------------------------------

.catch_warning <- function(expr, pattern = ".*") {
  msg <- NULL
  withCallingHandlers(expr,
    warning = function(w) {
      if (is.null(msg) && grepl(pattern, conditionMessage(w)))
        msg <<- conditionMessage(w)
      invokeRestart("muffleWarning")
    }
  )
  msg
}

.suppress_all <- function(expr) {
  suppressWarnings(expr)
}

# ---- Fixtures ----------------------------------------------------------------

set.seed(42)
pos_vals <- rexp(200, rate = 0.5)         # non-negative exponential
signed_vals <- rnorm(100)                 # signed
prob_vals   <- runif(100, 0, 1)           # probability bounded [0,1]
groups_2 <- rep(c("A", "B"), each = 100)

tmpdir <- tempfile("test_density_")
dir.create(tmpdir)
on.exit(unlink(tmpdir, recursive = TRUE), add = TRUE)

# ---- Unit: .detect_bounds ----------------------------------------------------

cat("1. .detect_bounds: non-negative -> left=0, right=NA...")
b <- .detect_bounds(c(0, 1.5, 3.0))
stopifnot(b$left == 0, is.na(b$right))
cat(" ok\n")

cat("2. .detect_bounds: probability -> left=0, right=1...")
b <- .detect_bounds(c(0, 0.3, 0.9, 1.0))
stopifnot(b$left == 0, b$right == 1)
cat(" ok\n")

cat("3. .detect_bounds: signed -> both NA...")
b <- .detect_bounds(c(-1.0, 0.5, 2.0))
stopifnot(is.na(b$left), is.na(b$right))
cat(" ok\n")

# ---- Unit: .group_colors -----------------------------------------------------

cat("4. .group_colors NULL cycles accent palette...")
cols <- .group_colors(NULL, c("A", "B", "C"))
stopifnot(cols[["A"]] == .ACCENT_COLORS[1])
stopifnot(cols[["B"]] == .ACCENT_COLORS[2])
stopifnot(length(cols) == 3)
cat(" ok\n")

cat("5. .group_colors single string -> uniform...")
cols <- .group_colors("red", c("A", "B"))
stopifnot(all(cols == "red"))
cat(" ok\n")

cat("6. .group_colors explicit vector...")
cols <- .group_colors(c("#aaa", "#bbb"), c("A", "B"))
stopifnot(cols[["A"]] == "#aaa", cols[["B"]] == "#bbb")
cat(" ok\n")

cat("7. .group_colors length mismatch -> error...")
err <- tryCatch(.group_colors(c("red", "blue", "green"), c("A", "B")), error = function(e) e)
stopifnot(inherits(err, "error"), grepl("color", conditionMessage(err)))
cat(" ok\n")

# ---- Unit: .reflect_kde ------------------------------------------------------

cat("8. .reflect_kde: zero outside left bound...")
d <- .reflect_kde(seq(0.01, 3.0, length.out = 100), bw = "nrd0", left = 0, right = NA_real_)
stopifnot(is.data.frame(d), all(c("x","y") %in% names(d)))
neg_density <- d$y[d$x < 0]
stopifnot(all(neg_density == 0))
stopifnot(max(d$y) > 0)
cat(" ok\n")

# ---- Integration: density mode -----------------------------------------------

cat("9. density single value -> gg object...")
p <- .suppress_all(pavlab_density(pos_vals))
stopifnot(inherits(p, "gg"))
cat(" ok\n")

cat("10. density writes file...")
outfile <- file.path(tmpdir, "dens.png")
.suppress_all(pavlab_density(pos_vals, filename = outfile))
stopifnot(file.exists(outfile), file.size(outfile) > 500)
cat(" ok\n")

cat("11. density warns about bounds for non-negative data...")
w <- .catch_warning(pavlab_density(pos_vals), "reflect_bounds|bounded")
stopifnot(!is.null(w))
cat(" ok\n")

cat("12. density no bounds warning for signed data...")
got_bounds_warn <- FALSE
withCallingHandlers(
  pavlab_density(signed_vals),
  warning = function(w) {
    if (grepl("reflect_bounds|bounded", conditionMessage(w))) got_bounds_warn <<- TRUE
    invokeRestart("muffleWarning")
  }
)
stopifnot(!got_bounds_warn)
cat(" ok\n")

cat("13. density reflect_bounds=TRUE: no density outside [0, inf)...")
p <- .suppress_all(pavlab_density(pos_vals, reflect_bounds = TRUE))
# The reflected KDE is plotted via geom_ribbon/geom_line with computed data;
# extract geom_line data from layer data via ggplot_build
built <- .suppress_all(ggplot2::ggplot_build(p))
line_data <- tryCatch(built$data[[2]], error = function(e) NULL)
if (!is.null(line_data) && "x" %in% names(line_data)) {
  neg_rows <- line_data[line_data$x < 0, ]
  if (nrow(neg_rows) > 0)
    stopifnot(all(abs(neg_rows$y) < 1e-6))
}
cat(" ok\n")

cat("14. density grouped overlay returns gg object...")
p <- .suppress_all(pavlab_density(pos_vals, groups = groups_2, kind = "density"))
stopifnot(inherits(p, "gg"))
cat(" ok\n")

cat("15. density warns with > 3 groups in overlay mode...")
g4 <- rep(c("A","B","C","D"), each = 25)
v4 <- rnorm(100)
w <- .catch_warning(
  pavlab_density(v4, groups = g4, kind = "density", facet = FALSE),
  "facet"
)
stopifnot(!is.null(w))
cat(" ok\n")

cat("16. density faceted: returns gg with facet_wrap layer...")
p <- .suppress_all(pavlab_density(pos_vals, groups = groups_2, kind = "density", facet = TRUE))
stopifnot(inherits(p, "gg"))
built <- .suppress_all(ggplot2::ggplot_build(p))
stopifnot(length(built$layout$layout$PANEL) == 2)
cat(" ok\n")

cat("17. density log_x label suffix...")
p <- .suppress_all(pavlab_density(pos_vals, log_x = "log2", xlabel = "TPM"))
xl <- p$labels$x
stopifnot(!is.null(xl), nchar(xl) > 0)
cat(" ok\n")

cat("18. density show_mean: vline layer present...")
p <- .suppress_all(pavlab_density(pos_vals, show_mean = TRUE))
layer_classes <- sapply(p$layers, function(l) class(l$geom)[1])
stopifnot(any(grepl("GeomVline", layer_classes)))
vl <- p$layers[[which(grepl("GeomVline", layer_classes))[1]]]
stopifnot(!is.null(vl$data$xintercept))
cat(" ok\n")

cat("19. density show_median: vline layer present...")
p <- .suppress_all(pavlab_density(pos_vals, show_median = TRUE))
layer_classes <- sapply(p$layers, function(l) class(l$geom)[1])
stopifnot(any(grepl("GeomVline", layer_classes)))
cat(" ok\n")

# ---- Integration: histogram mode ---------------------------------------------

cat("20. histogram returns gg object...")
p <- .suppress_all(pavlab_density(pos_vals, kind = "histogram"))
stopifnot(inherits(p, "gg"))
cat(" ok\n")

cat("21. histogram faceted returns gg with 2 panels...")
p <- .suppress_all(
  pavlab_density(pos_vals, groups = groups_2, kind = "histogram", facet = TRUE)
)
built <- .suppress_all(ggplot2::ggplot_build(p))
stopifnot(length(built$layout$layout$PANEL) == 2)
cat(" ok\n")

cat("22. histogram stat=count: geom_histogram with count aesthetic...")
p <- .suppress_all(pavlab_density(pos_vals, kind = "histogram", stat = "count"))
stopifnot(inherits(p, "gg"))
cat(" ok\n")

# ---- Integration: violin mode -----------------------------------------------

cat("23. violin requires groups -> error...")
err <- tryCatch(pavlab_density(pos_vals, kind = "violin"), error = function(e) e)
stopifnot(inherits(err, "error"), grepl("groups", conditionMessage(err)))
cat(" ok\n")

cat("24. violin returns gg object...")
p <- .suppress_all(pavlab_density(pos_vals, groups = groups_2, kind = "violin"))
stopifnot(inherits(p, "gg"))
cat(" ok\n")

cat("25. violin show_mean: hline layer present with correct yintercept...")
p <- pavlab_density(pos_vals, groups = groups_2, kind = "violin", show_mean = TRUE)
layer_classes <- sapply(p$layers, function(l) class(l$geom)[1])
stopifnot(any(grepl("GeomHline", layer_classes)))
hl <- p$layers[[which(grepl("GeomHline", layer_classes))[1]]]
expected <- mean(pos_vals)
stopifnot(abs(hl$data$yintercept - expected) < 1e-6)
cat(" ok\n")

cat("26. violin group order respected...")
p <- .suppress_all(
  pavlab_density(pos_vals, groups = groups_2, kind = "violin", order = c("B", "A"))
)
built <- .suppress_all(ggplot2::ggplot_build(p))
xlabels <- built$layout$panel_scales_x[[1]]$range$range
stopifnot(xlabels[1] == "B", xlabels[2] == "A")
cat(" ok\n")

# ---- Input validation -------------------------------------------------------

cat("27. unequal length -> error...")
err <- tryCatch(
  pavlab_density(pos_vals, groups = rep("A", length(pos_vals) - 1)),
  error = function(e) e
)
stopifnot(inherits(err, "error"), grepl("equal length|equal", conditionMessage(err)))
cat(" ok\n")

cat("28. nonfinite values dropped silently...")
vals_na <- c(1.0, NA, 2.0, Inf, 3.0)
p <- .suppress_all(pavlab_density(vals_na, kind = "histogram", stat = "count"))
built <- .suppress_all(ggplot2::ggplot_build(p))
hist_data <- built$data[[1]]
total_count <- sum(hist_data$count, na.rm = TRUE)
stopifnot(total_count == 3)
cat(" ok\n")

cat("29. theme: no top/right spines...")
p <- .suppress_all(pavlab_density(pos_vals))
built <- .suppress_all(ggplot2::ggplot_build(p))
theme_el <- ggplot2::theme_get()
p_theme  <- p$theme
stopifnot(inherits(p, "gg"))
cat(" ok\n")

cat("30. density single no-groups, reflect_bounds=TRUE produces geom_ribbon...")
p <- .suppress_all(pavlab_density(pos_vals, reflect_bounds = TRUE))
layer_classes <- sapply(p$layers, function(l) class(l$geom)[1])
stopifnot(any(grepl("GeomRibbon|GeomLine", layer_classes)))
cat(" ok\n")

cat("\nAll 30 tests passed.\n")

File: skills/plotting/tests/test_density.py
"""Tests for pavlab_density (Python/matplotlib).
Run from skills/plotting/:
    .venv/bin/pytest tests/test_density.py -v
"""

from __future__ import annotations

import sys
import warnings
from pathlib import Path

import numpy as np
import pytest

sys.path.insert(0, str(Path(__file__).resolve().parent.parent / "python"))

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import matplotlib.figure

from pavlab_density import _detect_bounds, _group_colors, _compute_kde, pavlab_density, _ACCENT_COLORS


# ---- fixtures ---------------------------------------------------------------

@pytest.fixture
def pos_vals():
    """200 positive exponential values."""
    return np.random.default_rng(0).exponential(2.0, 200)


@pytest.fixture
def grouped(pos_vals):
    groups = np.array(["A"] * 100 + ["B"] * 100)
    return pos_vals, groups


# ---- unit: helpers ----------------------------------------------------------

def test_detect_bounds_nonneg():
    lo, hi = _detect_bounds(np.array([0.0, 1.5, 3.0]))
    assert lo == 0.0
    assert hi is None


def test_detect_bounds_probability():
    lo, hi = _detect_bounds(np.array([0.0, 0.3, 0.9, 1.0]))
    assert lo == 0.0
    assert hi == 1.0


def test_detect_bounds_signed():
    lo, hi = _detect_bounds(np.array([-1.0, 0.5, 2.0]))
    assert lo is None
    assert hi is None


def test_group_colors_none_cycles_accent():
    colors = _group_colors(None, ["A", "B", "C"])
    assert colors[0] == _ACCENT_COLORS[0]
    assert colors[1] == _ACCENT_COLORS[1]
    assert len(colors) == 3


def test_group_colors_uniform_string():
    colors = _group_colors("red", ["A", "B"])
    assert all(c == "red" for c in colors)


def test_group_colors_explicit_list():
    colors = _group_colors(["#aaa", "#bbb"], ["A", "B"])
    assert colors == ["#aaa", "#bbb"]


def test_group_colors_length_mismatch():
    with pytest.raises(ValueError, match="color list length"):
        _group_colors(["red"], ["A", "B"])


def test_compute_kde_shape():
    arr = np.linspace(1, 5, 50)
    x, d = _compute_kde(arr, "scott")
    assert len(x) == len(d) == 512
    assert np.all(d >= 0)


def test_compute_kde_reflection_zeros_outside():
    arr = np.linspace(0.01, 3.0, 100)
    x, d = _compute_kde(arr, "scott", lo=0.0)
    assert d[x < 0].sum() == 0.0
    assert d.max() > 0


# ---- density mode -----------------------------------------------------------

def test_density_single_returns_axes(pos_vals):
    ax = pavlab_density(pos_vals)
    assert hasattr(ax, "collections") or hasattr(ax, "lines")
    plt.close("all")


def test_density_writes_file(pos_vals, tmp_path):
    out = tmp_path / "dens.png"
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        pavlab_density(pos_vals, filename=str(out))
    assert out.exists() and out.stat().st_size > 500
    plt.close("all")


def test_density_bounds_warning_when_nonneg(pos_vals):
    with pytest.warns(UserWarning, match="reflect_bounds"):
        pavlab_density(pos_vals)
    plt.close("all")


def test_density_no_bounds_warning_when_signed():
    vals = np.linspace(-3.0, 3.0, 100)
    with warnings.catch_warnings():
        warnings.simplefilter("error", UserWarning)
        try:
            pavlab_density(vals)
        except UserWarning as e:
            if "reflect_bounds" in str(e):
                pytest.fail("Should not warn about bounds for signed data")
    plt.close("all")


def test_density_reflect_bounds_no_leak():
    """With reflect_bounds=True, density curve should be zero for x < 0."""
    rng = np.random.default_rng(1)
    vals = rng.exponential(1.0, 300)
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        ax = pavlab_density(vals, reflect_bounds=True)
    lines = ax.get_lines()
    for ln in lines:
        xdata, ydata = ln.get_xdata(), ln.get_ydata()
        neg_mask = xdata < 0
        assert np.all(ydata[neg_mask] == pytest.approx(0.0, abs=1e-6)), \
            "density leaks below 0 despite reflect_bounds=True"
    plt.close("all")


def test_density_grouped_overlay_returns_axes(grouped):
    vals, groups = grouped
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        ax = pavlab_density(vals, groups=groups, kind="density")
    assert hasattr(ax, "lines")
    plt.close("all")


def test_density_grouped_warns_over_3_groups():
    rng = np.random.default_rng(2)
    vals = rng.standard_normal(100)
    groups = np.array(["A"] * 25 + ["B"] * 25 + ["C"] * 25 + ["D"] * 25)
    with pytest.warns(UserWarning, match="facet=True"):
        pavlab_density(vals, groups=groups, kind="density", facet=False)
    plt.close("all")


def test_density_faceted_returns_figure(grouped):
    vals, groups = grouped
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        result = pavlab_density(vals, groups=groups, kind="density", facet=True)
    assert isinstance(result, matplotlib.figure.Figure)
    assert len(result.axes) == 2
    plt.close("all")


def test_density_log_x_label_suffix(pos_vals):
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        ax = pavlab_density(pos_vals, log_x="log2", xlabel="TPM")
    assert "(log₂)" in ax.get_xlabel()
    plt.close("all")


def test_density_show_mean_vline(pos_vals):
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        ax = pavlab_density(pos_vals, show_mean=True)
    vlines = [ln for ln in ax.get_lines() if ln.get_linestyle() == "--"]
    assert len(vlines) >= 1
    plt.close("all")


def test_density_show_median_vline(pos_vals):
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        ax = pavlab_density(pos_vals, show_median=True)
    dotted = [ln for ln in ax.get_lines() if ln.get_linestyle() in (":", "dotted")]
    assert len(dotted) >= 1
    plt.close("all")


# ---- histogram mode ---------------------------------------------------------

def test_histogram_returns_axes(pos_vals):
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        ax = pavlab_density(pos_vals, kind="histogram")
    assert len(ax.patches) > 0
    plt.close("all")


def test_histogram_faceted_returns_figure(grouped):
    vals, groups = grouped
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        result = pavlab_density(vals, groups=groups, kind="histogram", facet=True)
    assert isinstance(result, matplotlib.figure.Figure)
    plt.close("all")


def test_histogram_shared_bins_grouped(grouped):
    """Faceted histogram panels should have identical bin edges."""
    vals, groups = grouped
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        fig = pavlab_density(vals, groups=groups, kind="histogram", facet=True)
    edges0 = [p.get_x() for p in fig.axes[0].patches]
    edges1 = [p.get_x() for p in fig.axes[1].patches]
    assert len(edges0) == len(edges1)
    plt.close("all")


# ---- violin mode ------------------------------------------------------------

def test_violin_requires_groups(pos_vals):
    with pytest.raises(ValueError, match="groups"):
        pavlab_density(pos_vals, kind="violin")


def test_violin_returns_axes(grouped):
    vals, groups = grouped
    ax = pavlab_density(vals, groups=groups, kind="violin")
    assert hasattr(ax, "collections")
    plt.close("all")


def test_violin_show_mean_hline(grouped):
    vals, groups = grouped
    ax = pavlab_density(vals, groups=groups, kind="violin", show_mean=True)
    hlines = [ln for ln in ax.get_lines() if ln.get_linestyle() == "--"]
    assert len(hlines) >= 1
    expected = float(np.mean(vals))
    y_vals = [float(ln.get_ydata()[0]) for ln in hlines]
    assert any(abs(y - expected) < 1e-6 for y in y_vals)
    plt.close("all")


def test_violin_group_order(grouped):
    vals, groups = grouped
    ax = pavlab_density(vals, groups=groups, kind="violin", order=["B", "A"])
    labels = [t.get_text() for t in ax.get_xticklabels()]
    assert labels == ["B", "A"]
    plt.close("all")


# ---- input validation -------------------------------------------------------

def test_unequal_length_raises(pos_vals):
    with pytest.raises(ValueError, match="equal length"):
        pavlab_density(pos_vals, groups=np.array(["A"] * (len(pos_vals) - 1)))


def test_nonfinite_dropped_silently():
    vals = np.array([1.0, np.nan, 2.0, np.inf, 3.0])
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        ax = pavlab_density(vals, kind="histogram", stat="count")
    total_count = sum(p.get_height() for p in ax.patches)
    assert total_count == pytest.approx(3.0)
    plt.close("all")


def test_theme_no_top_right_spine(pos_vals):
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        ax = pavlab_density(pos_vals)
    assert not ax.spines["top"].get_visible()
    assert not ax.spines["right"].get_visible()
    plt.close("all")
